src/vm/class.cpp

```diff
@@ -1346,33 +1345,30 @@ CorElementType EEClassLayoutInfo::GetNativeHFATypeRaw()
     {
         CorElementType fieldType = ELEMENT_TYPE_END;
 
-        switch (pFieldMarshaler->GetNStructFieldType())
-        {
-        case NFT_COPY4:
-        case NFT_COPY8:
-            fieldType = pFieldMarshaler->GetFieldDesc()->GetFieldType();
-            if (fieldType != ELEMENT_TYPE_R4 && fieldType != ELEMENT_TYPE_R8)
-                return ELEMENT_TYPE_END;
-            break;
-
-        case NFT_NESTEDLAYOUTCLASS:
-            fieldType = ((FieldMarshaler_NestedLayoutClass *)pFieldMarshaler)->GetMethodTable()->GetNativeHFAType();
-            break;
+        NativeFieldCategory category = pFieldMarshaler->GetNativeFieldCategory();
 
-        case NFT_NESTEDVALUECLASS:
-            fieldType = ((FieldMarshaler_NestedValueClass *)pFieldMarshaler)->GetMethodTable()->GetNativeHFAType();
-            break;
-
-        case NFT_FIXEDARRAY:
-            fieldType = ((FieldMarshaler_FixedArray *)pFieldMarshaler)->GetElementTypeHandle().GetMethodTable()->GetNativeHFAType();
-            break;
-
-        case NFT_DATE:
-            fieldType = ELEMENT_TYPE_R8;
-            break;
-
-        default:
-            // Not HFA
+        if ((category & NATIVE_FIELD_CATEGORY_FLOAT) == NATIVE_FIELD_CATEGORY_FLOAT)
```

Comment:
Feel free to remove the right-hand side of the `==`. I realize that contradicts my statement about types, but this is a common enough idiom that I think it is obvious how it works.